File: Javascript/2.addTwoNumber.js
O(max(m, n)) O(max(m, n))
var addTwoNumbers = function(l1, l2) {
    let result = new ListNode(0),
        node = result;
    while(l1 || l2) {
        let r = node.val,
            i = (l1 && l1.val) || 0,
            j = (l2 && l2.val) || 0,
            sum = r + i + j,
            m, n;
        if(sum >= 10) {
            m = 1;
            n = sum -10;
        }
        else {
            m = 0;
            n =sum;
        }
        l1 = l1 && l1.next;
        l2 = l2 && l2.next;
        node.val = n;
        if (m || l1 || l2) {
            node.next = new ListNode(m);
            node = node.next;
        }
    }
    return result;
}

File: Python/2.addTwoNumber.py
# O(max(m, n)) O(max(m, n))
class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        if not l1: return l2
        if not l2: return l1
        if l1.val + l2.val < 10:
            l3 = ListNode(l1.val + l2.val)
            l3.next = self.addTwoNumbers(l1.next, l2.next)
        else:
            l3 = ListNode(l1.val + l2.val - 10)
            temp = ListNode(1)
            temp.next = None
            l3.next = self.addTwoNumbers(l1.next, self.addTwoNumbers(l2.next, temp))
        return l3